onnxruntime/core/providers/cpu/ml/tree_ensemble_common.h

```diff
@@ -0,0 +1,706 @@
+};
+
+template <typename ITYPE, typename OTYPE>
+TreeEnsembleCommon<ITYPE, OTYPE>::TreeEnsembleCommon(int omp_tree, int omp_N) {
+  omp_tree_ = omp_tree;
+  omp_N_ = omp_N;
+  nodes_ = NULL;
+}
+
+template <typename ITYPE, typename OTYPE>
+TreeEnsembleCommon<ITYPE, OTYPE>::~TreeEnsembleCommon() {
+  if (nodes_ != NULL)
+    delete[] nodes_;
+}
+
+template <typename ITYPE, typename OTYPE>
+std::string TreeEnsembleCommon<ITYPE, OTYPE>::runtime_options() {
+  std::string res;
+#ifdef USE_OPENMP
+  res += "OPENMP";
+#endif
+  return res;
+}
+
+template <typename ITYPE, typename OTYPE>
+int TreeEnsembleCommon<ITYPE, OTYPE>::omp_get_max_threads() {
+#if USE_OPENMP
+  return ::omp_get_max_threads();
+#else
+  return 1;
+#endif
+}
+
+template <typename ITYPE, typename OTYPE>
+void TreeEnsembleCommon<ITYPE, OTYPE>::init(
+    const std::string& aggregate_function,
+    std::vector<OTYPE> base_values,
+    int64_t n_targets_or_classes,
+    std::vector<int64_t> nodes_falsenodeids,
+    std::vector<int64_t> nodes_featureids,
+    std::vector<OTYPE> nodes_hitrates,
+    std::vector<int64_t> nodes_missing_value_tracks_true,
+    const std::vector<std::string>& nodes_modes,
+    std::vector<int64_t> nodes_nodeids,
+    std::vector<int64_t> nodes_treeids,
+    std::vector<int64_t> nodes_truenodeids,
+    std::vector<OTYPE> nodes_values,
+    const std::string& post_transform,
+    std::vector<int64_t> target_class_ids,
+    std::vector<int64_t> target_class_nodeids,
+    std::vector<int64_t> target_class_treeids,
+    std::vector<OTYPE> target_class_weights) {
+  init_c(aggregate_function, base_values, n_targets_or_classes,
+         nodes_falsenodeids, nodes_featureids, nodes_hitrates,
+         nodes_missing_value_tracks_true, nodes_modes,
+         nodes_nodeids, nodes_treeids, nodes_truenodeids,
+         nodes_values, post_transform, target_class_ids,
+         target_class_nodeids, target_class_treeids,
+         target_class_weights);
+}
+
+template <typename ITYPE, typename OTYPE>
+void TreeEnsembleCommon<ITYPE, OTYPE>::init_c(
+    const std::string& aggregate_function,
+    const std::vector<OTYPE>& base_values,
+    int64_t n_targets_or_classes,
+    const std::vector<int64_t>& nodes_falsenodeids,
+    const std::vector<int64_t>& nodes_featureids,
+    const std::vector<OTYPE>& nodes_hitrates,
+    const std::vector<int64_t>& nodes_missing_value_tracks_true,
+    const std::vector<std::string>& nodes_modes,
+    const std::vector<int64_t>& nodes_nodeids,
+    const std::vector<int64_t>& nodes_treeids,
+    const std::vector<int64_t>& nodes_truenodeids,
+    const std::vector<OTYPE>& nodes_values,
+    const std::string& post_transform,
+    const std::vector<int64_t>& target_class_ids,
+    const std::vector<int64_t>& target_class_nodeids,
+    const std::vector<int64_t>& target_class_treeids,
+    const std::vector<OTYPE>& target_class_weights) {
+  aggregate_function_ = ::onnxruntime::ml::MakeAggregateFunction(aggregate_function);
```

Comment:
>::onnxruntime::ml:: [](start = 24, length = 19)

nit: is the '::onnxruntime::ml' qualifier required, and it so, is the '::' prefix needed?

Reply:
fixed